EIPS/eip-7480.md

```diff
@@ -0,0 +1,92 @@
+status: Draft
+type: Standards Track
+category: Core
+created: 2023-08-11
+requires: 3540, 3670
+---
+
+## Abstract
+
+Four new instrutions are introduced, that allow to read EOF container's data section: `DATALOAD` loads 32-byte word to stack, `DATALOADN` loads 32-byte word to stack where the word is addressed by a static immediate argument, `DATASIZE` loads data section size and `DATACOPY` copies a segment of data section to memory. 
+
+## Motivation
+
+Clear separation between code and data is one of the main features of EOF1. Data section may contain anything, e.g. compiler's metadata, but to make it useful for smart contracts, EVM has to have instructions that allow to read from data section. Previously existing instructions for bytecode inspection (`CODECOPY`, `CODESIZE` etc.) are deprecated in EOF1 and cannot be used for this purpose.
+
+`DATALOAD`, `DATASIZE`, `DATACOPY` instructions combination follows the design of existing instructions for reading other kinds of data (return data, call data).
+
+`DATALOADN` is an optimized version of `DATALOAD`, where data offset to read is set at compilation time, and therefore need not be validated at run-time, which makes the instruction cheaper.
+
+## Specification
+
+We introduce four new instructions on the same block number [EIP-3540](./eip-3540.md) is activated on:
+
+1 `DATALOAD` (0xe8)
+2.`DATALOADN` (0xe9)
+3.`DATASIZE` (0xea)
+4.`DATACOPY` (0xeb)
+
+If the code is legacy bytecode, all of these instructions result in an *exceptional halt*. (*Note: This means no change to behaviour.*)
+
+If the code is valid EOF1, the following execution rules apply:
+
+### `DATALOAD`
+
+1. Pops one value, `offset`, from the stack.
+2. If `offset + 32` is greater than the data section size of the active container, execution results in exceptional halt.
+3. Reads `[offset:offset+32]` segment from the data section and pushes it as 32-byte value to the stack.
+4. Deducts 3 gas.
+
+### `DATALOADN`
+
+1. Has one immediate argument,`offset`, encoded as a 16-bit unsigned big-endian value.
+2. Pops nothing from the stack.
+3. Reads `[offset:offset+32]` segment from the data section and pushes it as 32-byte value to the stack.
+4. Deducts 2 gas.
+
+`[offset:offset+32]` is guaranteed to be within data bounds by [code validation](#code-validation).
+
+### `DATASIZE`
+
+1. Pops nothing from the stack.
+2. Pushes the size of the data section of the active container to the stack.
+3. Deducts 2 gas.
+
+### `DATACOPY`
+
+1. Pops three values from the stack: `mem_offset`, `offset`, `size`.
+2. Performs memory expansion to `mem_offset + size` and deducts memory expansion cost.
+3. Deducts `3 * ((size + 31) // 32)` gas for copying.
+3. If `offset + size` is greater than data section size of the active container, execution results in exceptional halt.
+4. Reads `[offset:offset+size]` segment from the data section and writes it to memory starting at offset `mem_offset`.
```

Comment:
Shouldn't this also require validating `offset+size <= data_section_size`?

Reply:
This is in point 3 above?

Reply:
You're right